### PR TITLE
Remove unwarranted log.Fatal() statements, which can cause the Go runtime to panic if the YAML file does not already exist.

### PR DESCRIPTION
**Pull request by akramtexas**
- This PR is to address the following issue: [Issue #760](https://github.com/edgexfoundry/edgex-go/issues/760)
- The only reason why Go's built-in (overly-rudimentary) logging (viz. log.Fatal()) was used in the first place has to do with how the package structuring that was not allowing the import of the clients/logging package.
- Further testing revealed an issue, leading to removing those unwarranted log.Fatal() statements, primarily in the function FetchDockerComposeYamlAndPath(), which can cause the Go runtime to panic if the YAML file does not already exist.
- Also, the same function, FetchDockerComposeYamlAndPath(), contains a reference to an older YAML, as follows, which is being updated:
`composeUrl := "https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/compose-files/docker-compose-california-0.6.1.yml"`

### CODE REVIEW

File: internal/system/agent/executor/compose.go [hunk 1]
```diff
@@ -102,15 +101,14 @@ func RunDockerComposeCommand(service string, dockerComposeService string) {
...
-	// fmt.Sprintf("Pulling latest compose file from Github edgexfoundry repository...")
 	err = Do(func(attempt int) (bool, error) {
 		var err error
 		cmdDir, err = FetchDockerComposeYamlAndPath()
 		// Try 5 times
 		return attempt < 5, err
 	})
 	if err != nil {
-		log.Fatalln("Unable to pull the latest compose file from Github edgexfoundry repository.:", err)
+		log.Printf("Unable to pull the latest compose file from Github edgexfoundry repository: %s", err)
```
Comment (tsconn23) [inline internal/system/agent/executor/compose.go]: Any reason we're not using the same LoggingClient as the other services?

Comment (akramtexas) [reply internal/system/agent/executor/compose.go]: Yep, much as I would've preferred to use the _same_ LoggingClient as the other services, I ran into a circular-dependency issue. In fact, I alluded to that in the following point in the description atop this PR:
`The only reason why Go's built-in (overly-rudimentary) logging (viz. log.Fatal()) was used in the first place has to do with how the package structuring that was not allowing the import of the clients/logging package.`

File: internal/system/agent/executor/compose.go [hunk 2]
```diff
@@ -145,9 +140,9 @@ func FetchDockerComposeYamlAndPath() (string, error) {
 	body, _ := ioutil.ReadAll(res.Body)
 
 	// [1] Fetch contents of the latest "docker-compose.yml" file from Github.
-	resp, err := http.Get("https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/compose-files/docker-compose-california-0.6.0.yml")
+	resp, err := http.Get("https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/compose-files/docker-compose-delhi-0.7.0.yml")
```
[the remark this answers is not in the document]
Comment (akramtexas) [reply internal/system/agent/executor/compose.go]: I had intended to refactor this (Will do so, good catch) 👍

File: internal/system/agent/executor/compose.go (hunk 1 above)
Comment (akramtexas) [reply internal/system/agent/executor/compose.go]: This is the specific error seen on trying to use the same LoggingClient as the other services:
`CGO_ENABLED=0 go build -ldflags "-X github.com/edgexfoundry/edgex-go.Version=0.7.0" -o cmd/sys-mgmt-agent/sys-mgmt-agent ./cmd/sys-mgmt-agent
import cycle not allowed
package github.com/edgexfoundry/edgex-go/cmd/sys-mgmt-agent
        imports github.com/edgexfoundry/edgex-go/internal/system/agent
        imports github.com/edgexfoundry/edgex-go/internal/system/agent/executor
        imports github.com/edgexfoundry/edgex-go/internal/system/agent
make: *** [cmd/sys-mgmt-agent/sys-mgmt-agent] Error 1`

**Comment by akramtexas** [comment]
- Need to figure out how to install relevant cert(s) to fix the following (runtime) error:
> edgex-sys-mgmt   | 2018/11/08 18:37:55 Call to http.Get() failed with Get https://raw.githubusercontent.com/edgexfoundry/developer-scripts/master/compose-files/docker-compose-delhi-0.7.0.yml: x509: failed to load system roots and no roots provided

As @tsconn23 pointed out: 
`The link is "https" and it's unable to find the required certificate to make the secure connection`

File: internal/system/agent/executor/compose.go (hunk 1 above)
Comment (tsconn23) [reply internal/system/agent/executor/compose.go]: I just refactored the files in the executor package to use the EdgeX LoggingClient. I'm not seeing a circular import reference on build. You can find the work here.
https://github.com/tsconn23/edgex-go/tree/issue_763_tsc

File: internal/system/agent/executor/compose.go [hunk 3]
```diff
@@ -119,37 +118,27 @@ func RunDockerComposeCommand(service string, dockerComposeService string) {
 
 	out, err := cmd.CombinedOutput()
 	if err != nil {
-		log.Fatal(fmt.Sprintf("Call to docker-compose up -d failed with %s\n", err))
+		log.Printf(fmt.Sprintf("Call to docker-compose up -d failed with %s\n", err))
```
Comment (tsconn23) [inline internal/system/agent/executor/compose.go]: On statements like this, `err` should be `err.Error()`. The latter is a string which will replace `%s`. The former is not.

File: internal/system/agent/executor/compose.go [hunk 4]
```diff
@@ -119,37 +118,27 @@ func RunDockerComposeCommand(service string, dockerComposeService string) {
...
+		// agent.LoggingClient.Error("Call to docker-compose up -d failed with %s\n", err.Error())
 	}
-
-	// composeOutput := string(out)
-	// fmt.Sprintf("For the micro-service {%v}, we got this docker-compose output:\n {%v}", service, composeOutput)
 	println(out)
 
 	if ! WasDockerContainerComposeStarted(service) {
-		log.Fatal(fmt.Sprintf("The container for {%v} was NOT started!" + service))
+		log.Printf(fmt.Sprintf("The container for {%v} was NOT started!" + service))
```
Comment (tsconn23) [inline internal/system/agent/executor/compose.go]: `%v` should be `%s`. The `service` variable is a string type.

File: internal/system/agent/executor/compose.go (hunk 1 above)
Comment (akramtexas) [reply internal/system/agent/executor/compose.go]: **TODO:** Need to revisit in near-future (for standardized logging across services) 👍

**Comment by akramtexas** [comment]
Followed up on installing relevant cert(s) to fix the (runtime) error mentioned above. Thanking @JPWKU for helping out here, in modifying Dockerfile.sys-mgmt-agent like so:
```
RUN apk update && apk add make && apk add bash
RUN apk --no-cache add ca-certificates && update-ca-certificates
COPY . .
RUN make cmd/sys-mgmt-agent/sys-mgmt-agent

# FROM scratch
FROM golang:1.9-alpine
RUN apk update && apk add make && apk add bash
RUN apk --no-cache add ca-certificates && update-ca-certificates
```

**Comment by akramtexas** [comment]
- Note that there is no docker, no docker-compose binary available via SCRATCH.
- More fundamentally, however, a problem presents itself when running the SMA in-container, in that, unlike as originally anticipated (i.e. docker-compose command issued by the SMA would work across a network of Docker-containerized services seamlessly), it is problematic to have it operate (at this time) in a container. In other words, the SMA will be deployed natively to a given O/S, and thereafter issue docker-compose commands (as well as docker SDK commands), which will interface with and issue directives to Docker-containerized services, say, using the container.ID, etc.

**Comment by anonymouse64** [comment]
@akramtexas you probably need to run a Docker-in-Docker setup, with the docker.sock from the host mapped as a volume into the container and then your docker commands will work from inside the container. You may or may not need to also run the container as a privileged container. There's lots of articles walkthoughs online on how to run docker inside docker. Also note that docker-compose is merely a python package which acts as a wrapper around docker itself.

**Comment by akramtexas** [comment]
- At this point, I have tested (locally), with the SMA deployed natively to the O/S, and thereafter issuing docker-compose commands (as well as docker SDK commands), communicating with Docker-containerized services, these commands being triggered (behind-the-scenes) through invocations of the SMA's API.
- After wrapping up testing, I will squash commits, and update to that effect via a note here.

**Comment by akramtexas** [comment]
- As promised (in the prior comment, that, after wrapping up testing, I will squash commits, and update to that effect here), please know that that work is now _complete_. 
- Should I have missed anything, please let me know, thanks! @tsconn23 @jpwhitemn 
- Separately, I'll put together a brief _How-To guide_ for users of the system, to reflect this.

**Comment by tsconn23** [comment]
I am able to stop a service but not start one. If I stop edgex-core-data and then start it again, I see the following error in the log:

```
2018/11/09 09:56:39 Call to docker-compose up -d failed with exit status 1
[644/1536]0xc0003c9200
2018/11/09 09:57:06 The container for {%!!(MISSING)s(MISSING)} was NOT started!edgex-core-data
```

However the service returns the following via Postman
`Done. Started the requested services.`
The above message is returned because `InvokeOperation` always returns true and error conditions are not handled in `router.go::operationHandler`

My Postman request is a `POST` 
Url `http://localhost:48090/api/v1/operation`
Body `{"action":"start","services":["edgex-core-data"]}`

**Comment by anonymouse64** [comment]
@tsconn23 the bug I filed for not handling errors is https://github.com/edgexfoundry/edgex-go/issues/769

File: internal/system/agent/executor/compose.go (hunk 1 above)
Comment (tsconn23) [reply internal/system/agent/executor/compose.go]: I don't understand why this is a TODO when the code is in the branch referred to above and simply needs to be pulled in, but so be it.

Assigned to you https://github.com/edgexfoundry/edgex-go/issues/780

**Comment by anonymouse64** [comment]
I'm not totally clear how this is being tested, but I just wanted to point out 2 things about using docker with the `OperationType=os` configuration for the SMA:

* docker runs processes inside a pid namespace, which means that if the SMA runs inside of a docker container, it won't be able to see the processes that are running inside other containers. 
* on some linux distributions due to whatever configured Linux Security Modules (LSM's for short) the user might have, the ability for one to kill a process inside a pid namespace may be restricted. For example, on my Ubuntu 18.04 distribution using docker as a snap, which confines docker using AppArmor, Seccomp, etc. even as sudo I cannot kill a process running inside a docker container without using docker itself. This is due to how the LSM works where an unconfined thing (i.e. my root sudo shell) doesn't have full access to a confined thing (i.e. the process running inside a docker container), and I instead have to kill the process using something in the same confinement, i.e. docker itself. 

As such I highly discourage from using the `OperationType=os` in conjunction with SMA if the other services are running as docker containers. It's not totally clear to me if that's what y'all are doing in testing this PR, but just my 2¢

**Comment by akramtexas** [comment]
> I am able to stop a service but not start one. If I stop edgex-core-data and then start it again, I see the following error in the log:
> 
> ```
> 2018/11/09 09:56:39 Call to docker-compose up -d failed with exit status 1
> [644/1536]0xc0003c9200
> 2018/11/09 09:57:06 The container for {%!!(MISSING)s(MISSING)} was NOT started!edgex-core-data
> ```
> However the service returns the following via Postman
> `Done. Started the requested services.`
> The above message is returned because `InvokeOperation` always returns true and error conditions are not handled in `router.go::operationHandler`
> 
> My Postman request is a `POST`
> Url `http://localhost:48090/api/v1/operation`
> Body `{"action":"start","services":["edgex-core-data"]}`

- Thanks for the prompt feedback! I think the reason you're getting that error (i.e. "...The container for {%!!(MISSING)v(MISSING)} was NOT started!edgex-core-data...") is that you're not launching the SMA from the following level:
/Users/<your-username>/go/src/github.com/edgexfoundry/edgex-go/cmd/sys-mgmt-agent

**Comment by akramtexas** [comment]
> I'm not totally clear how this is being tested, but I just wanted to point out 2 things about using docker with the `OperationType=os` configuration for the SMA:
> 
> * docker runs processes inside a pid namespace, which means that if the SMA runs inside of a docker container, it won't be able to see the processes that are running inside other containers (and as such needs to use the `docker` commands to control those other services).
> * on some linux distributions due to whatever configured Linux Security Modules (LSM's for short) the user might have, the ability for one to kill a process inside a pid namespace may be restricted. For example, on my Ubuntu 18.04 distribution using docker as a snap, which confines docker using AppArmor, Seccomp, etc. even as sudo I cannot kill a process running inside a docker container without using docker itself. This is due to how the LSM works where an unconfined thing (i.e. my root shell) doesn't have full access to a confined thing (i.e. the process running inside a docker container), and I instead have to kill the process using something in the same confinement, i.e. docker itself.
> 
> As such I highly discourage from using the `OperationType=os` in conjunction with SMA if the other services are running as docker containers. It's not totally clear to me if that's what y'all are doing in testing this PR, but just my 2¢

- Thanks for making the time to share the detailed feedback, appreciated!
- To help you with gaining insight into testing the SMA capabilities, I invite you to please check out a brief (10-minute) YouTub video I've put together in the past half-hour. This walkthrough showing the capabilities offered by the SMA. It's a bit developer-centric, and is now available here: [SMA (aka "System Management Agent") Demo.](https://youtu.be/tlEnEo6whas)

**Comment by akramtexas** [comment]
> This PR has segued away from addressing logging to questions regarding basic functionality (logging --> docker compose --> HTTPS certs --> non-Dockerized execution --> false positives). At this point, for me, the criteria to approve this PR has grown to validate that all basic functionality committed to for the Delhi release works to the satisfaction of everyone involved. Based on comments above, we're not there yet. Do we need a call to discuss?

- Totally valid point there: This PR _has_ segued away from addressing logging to questions regarding basic functionality, and, correspondingly, the criteria to approve this PR have grown. Let's connect soon to discuss.

**Comment by anonymouse64** [comment]
@akramtexas thanks for the video, but it doesn't answer my questions about how folks are testing this PR locally... For clarity, I am wondering the following specific things:

1. When someone is deploying EdgeX and they use the docker-compose script, will `sys-mgmt-agent` be run inside a docker container?
2. If the `sys-mgmt-agent` is run inside a docker container and has the value of `"docker"` for `OperationType` in the configuration.toml for `sys-mgmt-agent`, how will it communicate with the docker socket running on the host?
3. If someone is running EdgeX using the docker-compose script, what is the value they should use for `OperationType` in the configuration.toml for `sys-mgmt-agent`?

**Comment by akramtexas** [comment]
> @akramtexas thanks for the video, but it doesn't answer my questions about how folks are testing this PR locally... For clarity, I am wondering the following specific things:
> 
> 1. When someone is deploying EdgeX and they use the docker-compose script, will `sys-mgmt-agent` be run inside a docker container?
> 2. If the `sys-mgmt-agent` is run inside a docker container and has the value of `"docker"` for `OperationType` in the configuration.toml for `sys-mgmt-agent`, how will it communicate with the docker socket running on the host?
> 3. If someone is running EdgeX using the docker-compose script, what is the value they should use for `OperationType` in the configuration.toml for `sys-mgmt-agent`?

- Sure thing @anonymouse64 👍 

1. So (1) the entry for the sys-mgmt-agent service has been removed from the Makefile (i.e. "github.com/edgexfoundry/edgex-go/Makefile"), and (2) the sys-mgmt-agent service's Dockerfile itself (i.e. "Dockerfile.sys-mgmt-agent") has been deleted from the project (on this branch of my fork of the edgex-go project). The relevant entry (for the sys-mgmt-agent service image) _needs_ to be removed from docker-compose.yml.
2. I am contemplating, pending this ongoing discussion (and at least _per_ this PR, _for_ this Delhi Dot Release) to have the SMA run natively outside of any container.

**Comment by tsconn23** [comment]
@akramtexas  wrote:
> * Thanks for the prompt feedback! I think the reason you're getting that error (i.e. "...The container for {%!!(MISSING)v(MISSING)} was NOT started!edgex-core-data...") is that you're not launching the SMA from the following directory level:
>   /Users//go/src/github.com/edgexfoundry/edgex-go/cmd/sys-mgmt-agent

But I AM launching it from there
```
**~/go/src/github.com/edgexfoundry/edgex-go/cmd/sys-mgmt-agent $ ./sys-mgmt-agent**
INFO: 2018/11/09 13:06:43 Service dependencies resolved...
INFO: 2018/11/09 13:06:43 Starting edgex-sys-mgmt-agent 0.7.0 
INFO: 2018/11/09 13:06:43 This is the System Management Agent Service
INFO: 2018/11/09 13:06:43 Service started in: 2.920551ms
INFO: 2018/11/09 13:06:43 Listening on port: 48090
INFO: 2018/11/09 13:06:58 About to {stop} the service {edgex-core-data} with parameters {[graceful]}.
[13/13]0xc0004e8000
[12/13]0xc0001cec00
ERROR: 2018/11/09 13:07:02 edgex-core-data get metrics error: Get http://localhost:48080/api/v1/metrics: dial tcp [::1]:48080: connect: connection refused
INFO: 2018/11/09 13:07:19 About to {start} the service {edgex-core-data} with parameters {[]}.
2018/11/09 13:07:35 Call to docker-compose up -d failed with exit status 1
[228/1536]0xc000620600
2018/11/09 13:07:35 The container for {%!!(MISSING)s(MISSING)} was NOT started!edgex-core-data
```
As described, response from Postman is stil as follows:
`Done. Started the requested services.`

**Comment by anonymouse64** [comment]
@akramtexas I think your points merit further discussion on a separate issue because I'm worried about how a user will be deploying this in the docker case - if as you mention there is no docker container for the `sys-mgmt-agent`, how will they be expected to deploy the service? Will they have to download or build the binaries and run them manually? Anyways, I think discussion of SMA deployment for Delhi should move to a proper issue so I filed one here: https://github.com/edgexfoundry/edgex-go/issues/781

**Comment by akramtexas** [comment]
- Thanking @tsconn23 and @anonymouse64 for working closely with me on this one.
- Separately, I will _now_ update **master** branch correspondingly (This PR was for the **delhi** branch).

**Pull request merged**